### PR TITLE
Support uncurried higher order functions

### PR DESCRIPTION
**Pull request by paulcadman**
### Higher order function types

A function type argument is translated to a C function pointer

For example (see: https://github.com/heliaxdev/minijuvix/blob/47d9204b4c6391c635320ffb72591dba915e8e72/tests/positive/MiniC/HigherOrder/Input.mjuvix#L133)

```
foldl : (Nat → Nat → Nat) → Nat → ListNat → Nat;
```

is translated to:

```
Nat_9_t * foldl_26(Nat_9_t * (* fa0)(Nat_9_t *, Nat_9_t *),
                   Nat_9_t * fa1,
                   ListNat_23_t * fa2)
```

### Passing higher-order functions

Function pointers can be passed directly as arguments. In the definition of `sum` (see https://github.com/heliaxdev/minijuvix/blob/47d9204b4c6391c635320ffb72591dba915e8e72/tests/positive/MiniC/HigherOrder/Input.mjuvix#L141)

```
infixl 6 +;
+ : Nat → Nat → Nat;
+ zero n ≔ n;
+ (suc m) n ≔ suc (m + n);

sum : ListNat → Nat;
sum l ≔ foldl (+) zero l;
```

The plus function is translated with signature:

```
Nat_9_t * v_12(Nat_9_t * fa0, Nat_9_t * fa1)
```

And the sum function is translated to:

```
Nat_9_t * sum_33(ListNat_23_t * fa0)
{
    return foldl_26(v_12, new_Zero_10, fa0);
}
```

### Translating nullary functions

The change that makes passing higher-order functions work is at https://github.com/heliaxdev/minijuvix/compare/90-c-hof?expand=1#diff-ef2e5040dfaf26feb9a1737f0e280c882f9181f4c3f9664510673af1b981f862R297

That is, previously `IdenFunction`s were translated to nullary function calls. This doesn't work now that `IdenFunction`s can be higher-order functions that need to be passed by name as C function pointers.

The solution here is to wrap all nullary function calls (constructors and top-level functions) in a macro so they can be referred to without a function call.

For example in the translation of:

```
inductive Nat {
  zero : Nat;
  suc : Nat → Nat;
};
```

The zero constructor is translated to a nullary C function:

```
static inline Nat_13_t * new_Zero_14_nullary()
```

and a macro:

```
#define new_Zero_14 (new_Zero_14_nullary())
```
so that it can be referenced subsequently by the unapplied name`new_Zero_14`.

### CODE REVIEW

**Review by janmasrovira — CHANGES_REQUESTED** [review]
Looks great! I just have some small suggestions

File: src/MiniJuvix/Syntax/MiniC/Serialization.hs [hunk 1]
```diff
@@ -33,19 +39,61 @@ serialize = show . codeUnitDoc
...
+      C.undefNode
+    where
+      declrName :: CDeclr
+      declrName = CDeclr (mkIdent <$> _declName) ptrDeclr Nothing [] C.undefNode
+      ptrDeclr :: [CDerivedDeclarator C.NodeInfo]
+      ptrDeclr = [CPtrDeclr [] C.undefNode | _declIsPtr]
+      initializer :: Maybe CInit
+      initializer = mkCInit <$> _declInitializer
+
+-- FuncPointerTypeDecl (FuncPointerType {..}) ->
```
[the remark this answers is not in the document]
Comment (paulcadman) [reply src/MiniJuvix/Syntax/MiniC/Serialization.hs:82]: No - I meant to delete this, thanks

File: src/MiniJuvix/Translation/MonoJuvixToMiniC.hs [hunk 2]
```diff
@@ -172,16 +179,35 @@ goFunctionDef Mono.FunctionDef {..} =
                   }
               )
           )
+
+    isNullary :: Bool
+    isNullary = null funArgTypes && funcBasename /= main_
+    funcBasename :: Text
+    funcBasename = mkName _funDefName
+    nullaryDefine :: [Define]
```
[the remark this answers is not in the document]
Comment (paulcadman) [reply src/MiniJuvix/Translation/MonoJuvixToMiniC.hs:187]: Agreed

**Review by janmasrovira — APPROVED** [review]
Great!

**Pull request merged**